File: solutions/union-find/547-number-of-provinces.py
"""
547. Number of Provinces
Medium

There are n cities. Some of them are connected, while some are not. If city a is
connected directly with city b, and city b is connected directly with city c,
then city a is connected indirectly with city c.

A province is a group of directly or indirectly connected cities and no other
cities outside of the group.

You are given an n x n matrix isConnected where isConnected[i][j] = 1 if the ith
city and the jth city are directly connected, and isConnected[i][j] = 0 otherwise.

Return the total number of provinces.

Example:
Input: isConnected = [[1,1,0],[1,1,0],[0,0,1]]
Output: 2
"""

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
This is a classic connected components problem. We need to find groups of cities that are directly or indirectly connected. Each connected component represents one province.

### APPROACH:
1. **Union-Find (Disjoint Set)**: Perfect for tracking connected components
2. **Initialize**: Each city starts as its own province
3. **Union Operations**: For each connection, union the two cities
4. **Count Components**: Count the number of distinct root parents

### WHY THIS WORKS:
Union-Find efficiently maintains connected components. When we union two cities, we're merging their provinces. The number of unique root parents at the end equals the number of provinces.

### TIME COMPLEXITY: O(n² × α(n))
- n² to iterate through the matrix
- α(n) is the inverse Ackermann function (practically constant)
- With path compression and union by rank: nearly O(n²)

### SPACE COMPLEXITY: O(n)
- Union-Find parent and rank arrays

### EXAMPLE WALKTHROUGH:
For isConnected = [[1,1,0],[1,1,0],[0,0,1]]:
```
Cities: 0, 1, 2
Connections: 0-1 (cities 0 and 1 connected)
Initial: parent = [0, 1, 2], components = 3
Union(0,1): parent = [0, 0, 2], components = 2
Final: 2 provinces (cities {0,1} and {2})
```

### ALTERNATIVE APPROACHES:
1. **DFS**: Visit all connected cities starting from each unvisited city
2. **BFS**: Similar to DFS but using queue instead of recursion
3. **Union-Find**: Most efficient for this specific problem type

### EDGE CASES:
- Single city: return 1
- No connections: each city is its own province
- All connected: return 1
- Self-connections (diagonal): already handled in matrix

### UNION-FIND OPTIMIZATIONS:
- **Path Compression**: Make nodes point directly to root
- **Union by Rank**: Attach smaller tree under root of larger tree
- Together these make operations nearly O(1) amortized

</details>

class Solution:
    def findCircleNum(self, isConnected: list[list[int]]) -> int:
        """
        Approach: Union Find
        Time Complexity: O(n² * α(n)) where α is inverse Ackermann
        Space Complexity: O(n)
        """
        n = len(isConnected)
        parent = list(range(n))
        rank = [0] * n

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])  # Path compression
            return parent[x]

        def union(x, y):
            px, py = find(x), find(y)
            if px == py:
                return

            # Union by rank
            if rank[px] < rank[py]:
                parent[px] = py
            elif rank[px] > rank[py]:
                parent[py] = px
            else:
                parent[py] = px
                rank[px] += 1

        # Build union find
        for i in range(n):
            for j in range(i + 1, n):
                if isConnected[i][j] == 1:
                    union(i, j)

        # Count unique provinces
        return len(set(find(i) for i in range(n)))

    def findCircleNumDFS(self, isConnected: list[list[int]]) -> int:
        """
        Approach: DFS
        Time Complexity: O(n²)
        Space Complexity: O(n)
        """
        n = len(isConnected)
        visited = [False] * n
        provinces = 0

        def dfs(city):
            for neighbor in range(n):
                if isConnected[city][neighbor] == 1 and not visited[neighbor]:
                    visited[neighbor] = True
                    dfs(neighbor)

        for city in range(n):
            if not visited[city]:
                visited[city] = True
                dfs(city)
                provinces += 1

        return provinces


"""
684. Redundant Connection
Medium

In this problem, a tree is an undirected graph that is connected and has no cycles.

You are given a graph that started as a tree with n nodes labeled from 1 to n,
with one additional edge added.

Return an edge that can be removed so that the resulting graph is a tree of n nodes.

Example:
Input: edges = [[1,2],[1,3],[2,3]]
Output: [2,3]
"""

class SolutionRedundant:
    def findRedundantConnection(self, edges: list[list[int]]) -> list[int]:
        """
        Approach: Union Find
        Time Complexity: O(n * α(n))
        Space Complexity: O(n)
        """
        parent = {}

        def find(x):
            if x not in parent:
                parent[x] = x
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            px, py = find(x), find(y)
            if px == py:
                return False  # Cycle detected
            parent[px] = py
            return True

        for u, v in edges:
            if not union(u, v):
                return [u, v]

        return []


"""
721. Accounts Merge
Medium

Given a list of accounts where each element accounts[i] is a list of strings,
where the first element accounts[i][0] is a name, and the rest of the elements
are emails representing emails of the account.

Merge the accounts and return the accounts in sorted order.

Example:
Input: accounts = [["John","johnsmith@mail.com","john_newyork@mail.com"],
                   ["John","johnsmith@mail.com","john00@mail.com"],
                   ["Mary","mary@mail.com"],
                   ["John","johnnybravo@mail.com"]]
Output: [["John","john00@mail.com","john_newyork@mail.com","johnsmith@mail.com"],
         ["Mary","mary@mail.com"],
         ["John","johnnybravo@mail.com"]]
"""

class SolutionAccounts:
    def accountsMerge(self, accounts: list[list[str]]) -> list[list[str]]:
        """
        Approach: Union Find with email mapping
        Time Complexity: O(n * k * α(n)) where k is average emails per account
        Space Complexity: O(n * k)
        """
        from collections import defaultdict

        parent = {}
        email_to_name = {}

        def find(x):
            if x not in parent:
                parent[x] = x
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            parent[find(x)] = find(y)

        # Process each account
        for account in accounts:
            name = account[0]
            first_email = account[1]

            for email in account[1:]:
                email_to_name[email] = name
                union(first_email, email)

        # Group emails by component
        components = defaultdict(list)
        for email in email_to_name:
            components[find(email)].append(email)

        # Build result
        result = []
        for emails in components.values():
            name = email_to_name[emails[0]]
            result.append([name] + sorted(emails))

        return result


# Test cases
if __name__ == "__main__":
    # Test Number of Provinces
    solution = Solution()

    print("Number of Provinces:")
    test_cases = [
        [[1, 1, 0], [1, 1, 0], [0, 0, 1]],
        [[1, 0, 0], [0, 1, 0], [0, 0, 1]],
        [[1, 1, 1], [1, 1, 1], [1, 1, 1]]
    ]

    for matrix in test_cases:
        result = solution.findCircleNum(matrix)
        print(f"Connections: {matrix}")
        print(f"Provinces: {result}\n")

    # Test Redundant Connection
    solution_redundant = SolutionRedundant()

    print("Redundant Connection:")
    edges_cases = [
        [[1, 2], [1, 3], [2, 3]],
        [[1, 2], [2, 3], [3, 4], [1, 4], [1, 5]]
    ]

    for edges in edges_cases:
        result = solution_redundant.findRedundantConnection(edges)
        print(f"Edges: {edges}")
        print(f"Redundant: {result}\n")

File: solutions/union-find/alternatives/547-number-of-provinces.js
/**
 * Difficulty: Medium
 *
 * [Problem description goes here]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input description]</dd>
 * <dt>Output:</dt>
 * <dd>[output description]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [High-level insight or key observation]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach]
 *
 * ### WHY THIS WORKS:
 * - [Explanation of correctness]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [example input]
 * ```
 *
 * **Step 1:** [description]
 *
 * **Step 2:** [description]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### EDGE CASES:
 * - **[Edge case 1]:** [how it's handled]
 * - **[Edge case 2]:** [how it's handled]
 *
 * </details>
 */

/**
 * Union-Find data structure for tracking connected components
 */
class UnionFind {
    constructor(n) {
        this.parent = Array.from({length: n}, (_, i) => i);
        this.rank = new Array(n).fill(0);
        this.components = n;
    }

    /**
     * Find root with path compression
     * @param {number} x
     * @returns {number} Root of x
     */
    find(x) {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }

    /**
     * Union by rank. Returns true if union occurred
     * @param {number} x
     * @param {number} y
     * @returns {boolean} True if union occurred
     */
    union(x, y) {
        const rootX = this.find(x);
        const rootY = this.find(y);

        if (rootX === rootY) {
            return false; // Already connected
        }

        // Union by rank
        if (this.rank[rootX] < this.rank[rootY]) {
            this.parent[rootX] = rootY;
        } else if (this.rank[rootX] > this.rank[rootY]) {
            this.parent[rootY] = rootX;
        } else {
            this.parent[rootY] = rootX;
            this.rank[rootX]++;
        }

        this.components--;
        return true;
    }
}

/**
 * Find number of provinces using Union-Find
 * @param {number[][]} isConnected - n x n adjacency matrix where isConnected[i][j] = 1 means city i and city j are directly connected
 * @returns {number} Number of provinces (connected components)
 *
 * Time Complexity: O(n² × α(n)) where α is inverse Ackermann
 * Space Complexity: O(n) for Union-Find structure
 */
function findCircleNum(isConnected) {
    const n = isConnected.length;
    const uf = new UnionFind(n);

    // Process all connections
    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) { // Only check upper triangle (symmetric matrix)
            if (isConnected[i][j] === 1) {
                uf.union(i, j);
            }
        }
    }

    return uf.components;
}

/**
 * Alternative DFS solution
 * @param {number[][]} isConnected - Adjacency matrix
 * @returns {number} Number of provinces
 *
 * Time Complexity: O(n²)
 * Space Complexity: O(n) for visited array and recursion stack
 */
function findCircleNumDFS(isConnected) {
    const n = isConnected.length;
    const visited = new Array(n).fill(false);
    let provinces = 0;

    /**
     * Mark all cities in current province as visited
     * @param {number} city
     */
    function dfs(city) {
        visited[city] = true;
        for (let neighbor = 0; neighbor < n; neighbor++) {
            if (isConnected[city][neighbor] === 1 && !visited[neighbor]) {
                dfs(neighbor);
            }
        }
    }

    // Find all provinces
    for (let i = 0; i < n; i++) {
        if (!visited[i]) {
            // Found a new province
            dfs(i);
            provinces++;
        }
    }

    return provinces;
}

/**
 * Alternative BFS solution
 * @param {number[][]} isConnected - Adjacency matrix
 * @returns {number} Number of provinces
 */
function findCircleNumBFS(isConnected) {
    const n = isConnected.length;
    const visited = new Array(n).fill(false);
    let provinces = 0;

    for (let i = 0; i < n; i++) {
        if (!visited[i]) {
            // Start BFS for new province
            const queue = [i];
            visited[i] = true;

            while (queue.length > 0) {
                const city = queue.shift();
                for (let neighbor = 0; neighbor < n; neighbor++) {
                    if (isConnected[city][neighbor] === 1 && !visited[neighbor]) {
                        visited[neighbor] = true;
                        queue.push(neighbor);
                    }
                }
            }

            provinces++;
        }
    }

    return provinces;
}

/**
 * Test cases for Problem 547: Number of Provinces
 */
function testSolution() {
    console.log('Testing 547. Number of Provinces');

    // Test case 1: Two provinces
    const isConnected1 = [[1,1,0],[1,1,0],[0,0,1]];
    const result1 = findCircleNum(isConnected1);
    const expected1 = 2;
    console.assert(result1 === expected1, `Test 1 failed: expected ${expected1}, got ${result1}`);

    // Test case 2: Three provinces (all isolated)
    const isConnected2 = [[1,0,0],[0,1,0],[0,0,1]];
    const result2 = findCircleNum(isConnected2);
    const expected2 = 3;
    console.assert(result2 === expected2, `Test 2 failed: expected ${expected2}, got ${result2}`);

    // Test case 3: One province (all connected)
    const isConnected3 = [[1,1,1],[1,1,1],[1,1,1]];
    const result3 = findCircleNum(isConnected3);
    const expected3 = 1;
    console.assert(result3 === expected3, `Test 3 failed: expected ${expected3}, got ${result3}`);

    // Test case 4: Single city
    const isConnected4 = [[1]];
    const result4 = findCircleNum(isConnected4);
    const expected4 = 1;
    console.assert(result4 === expected4, `Test 4 failed: expected ${expected4}, got ${result4}`);

    // Test case 5: Complex connections
    const isConnected5 = [[1,0,0,1],[0,1,1,0],[0,1,1,1],[1,0,1,1]];
    const result5 = findCircleNum(isConnected5);
    const expected5 = 1; // All cities are connected through paths
    console.assert(result5 === expected5, `Test 5 failed: expected ${expected5}, got ${result5}`);

    // Test DFS approach
    const result6 = findCircleNumDFS(isConnected1);
    console.assert(result6 === expected1, `DFS test failed: expected ${expected1}, got ${result6}`);

    // Test BFS approach
    const result7 = findCircleNumBFS(isConnected1);
    console.assert(result7 === expected1, `BFS test failed: expected ${expected1}, got ${result7}`);

    console.log('All test cases passed for 547. Number of Provinces!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 547. Number of Provinces ===');
    console.log('Category: Union Find');
    console.log('Difficulty: Medium');
    console.log('');

    // Example 1: Two provinces
    console.log('Example 1: Two provinces');
    const isConnected1 = [[1,1,0],[1,1,0],[0,0,1]];
    const result1 = findCircleNum(isConnected1);
    console.log(`findCircleNum(${JSON.stringify(isConnected1)}) -> ${result1}`);
    console.log('Explanation: Cities 0 and 1 form one province, city 2 forms another\n');

    // Example 2: Three isolated cities
    console.log('Example 2: Three isolated cities');
    const isConnected2 = [[1,0,0],[0,1,0],[0,0,1]];
    const result2 = findCircleNum(isConnected2);
    console.log(`findCircleNum(${JSON.stringify(isConnected2)}) -> ${result2}`);
    console.log('Explanation: Each city is its own province\n');

    // Example 3: All connected
    console.log('Example 3: All connected');
    const isConnected3 = [[1,1,1],[1,1,1],[1,1,1]];
    const result3 = findCircleNum(isConnected3);
    console.log(`findCircleNum(${JSON.stringify(isConnected3)}) -> ${result3}`);
    console.log('Explanation: All cities form one big province\n');

    // Compare different approaches
    console.log('Algorithm comparison:');
    const approaches = [
        ['Union-Find', findCircleNum],
        ['DFS', findCircleNumDFS],
        ['BFS', findCircleNumBFS]
    ];

    for (const [name, method] of approaches) {
        const result = method(isConnected1);
        console.log(`${name}: ${result}`);
    }

    console.log('\nKey insights:');
    console.log('1. Connected components problem - find separate groups');
    console.log('2. Union-Find efficiently manages component membership');
    console.log('3. DFS/BFS can mark all cities in a component as visited');
    console.log('4. Matrix is symmetric (undirected graph)');
    console.log('5. Count unique components to get number of provinces');

    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    findCircleNum,
    findCircleNumDFS,
    findCircleNumBFS,
    UnionFind,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution uses Union-Find with path compression and union by rank
 * - Time complexity is nearly O(n²) due to inverse Ackermann function
 * - Alternative DFS/BFS approaches also solve in O(n²) time
 * - The problem reduces to counting connected components in an undirected graph
 * - Essential insight: each connected component represents one province
 * - Union-Find provides optimal solution for dynamic connectivity queries
 */
